File: src/runtime.ts
import { callCapability, listCapabilities } from "./capabilities.js";
import { log } from "./logger.js";
import type { PluginManifest, JsonObject, JsonValue } from "./types.js";
import { RpcPeer } from "./rpc.js";

export class PluginRuntime {
  constructor(
    private manifest: PluginManifest,
    private peer: RpcPeer,
  ) {}

  registerHandlers(): void {
    this.peer.on("plugin.handshake", () => this.handshake());
    this.peer.on("plugin.start", () => this.start());
    this.peer.on("plugin.stop", () => this.stop());
    this.peer.on("plugin.health", () => this.health());
    this.peer.on("capability.call", (params) => this.capabilityCall(params));
  }

  async listen(): Promise<void> {
    this.registerHandlers();
    log("plugin ready", { id: this.manifest.id });
    await this.peer.listen();
  }

  private handshake(): JsonObject {
    return {
      pluginId: this.manifest.id,
      protocolVersion: "0.1",
      capabilities: this.manifest.capabilities?.map((capability) => capability.name) ?? [],
    };
  }

  private start(): JsonObject {
    log("plugin started", { id: this.manifest.id });
    return { started: true };
  }

  private stop(): JsonObject {
    log("plugin stopped", { id: this.manifest.id });
    process.exitCode = 0;
    return { stopped: true };
  }

  private health(): JsonObject {
    return {
      ok: true,
      pluginId: this.manifest.id,
      capabilities: listCapabilities(),
    };
  }

  private async capabilityCall(params: JsonValue | undefined): Promise<JsonValue> {
    if (!isObject(params) || typeof params.name !== "string") {
      throw new Error("capability.call requires params.name");
    }

    return callCapability(params.name, params.input);
  }
}

function isObject(value: JsonValue | undefined): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: templates/go/internal/plugin/runtime.go
package plugin

import (
	"encoding/json"
	"fmt"
	"io"
)

type Runtime struct {
	manifest Manifest
	rpc      *RPC
	logs     io.Writer
}

func NewRuntime(manifestPath string, input io.Reader, output io.Writer, logs io.Writer) (*Runtime, error) {
	manifest, err := LoadManifest(manifestPath)
	if err != nil {
		return nil, err
	}

	r := &Runtime{
		manifest: manifest,
		rpc:      NewRPC(input, output),
		logs:     logs,
	}
	r.registerHandlers()
	return r, nil
}

func (r *Runtime) Listen() error {
	fmt.Fprintf(r.logs, "[auxitalk-plugin] plugin ready id=%s\n", r.manifest.ID)
	return r.rpc.Listen()
}

func (r *Runtime) registerHandlers() {
	r.rpc.Handle("plugin.handshake", r.handshake)
	r.rpc.Handle("plugin.start", r.start)
	r.rpc.Handle("plugin.stop", r.stop)
	r.rpc.Handle("plugin.health", r.health)
	r.rpc.Handle("capability.call", r.capabilityCall)
}

func (r *Runtime) handshake(_ json.RawMessage) (any, error) {
	caps := make([]string, 0, len(r.manifest.Capabilities))
	for _, cap := range r.manifest.Capabilities {
		caps = append(caps, cap.Name)
	}
	return map[string]any{
		"pluginId":        r.manifest.ID,
		"protocolVersion": "0.1",
		"capabilities":     caps,
	}, nil
}

func (r *Runtime) start(_ json.RawMessage) (any, error) {
	fmt.Fprintf(r.logs, "[auxitalk-plugin] plugin started id=%s\n", r.manifest.ID)
	return map[string]any{"started": true}, nil
}

func (r *Runtime) stop(_ json.RawMessage) (any, error) {
	fmt.Fprintf(r.logs, "[auxitalk-plugin] plugin stopped id=%s\n", r.manifest.ID)
	return map[string]any{"stopped": true}, nil
}

func (r *Runtime) health(_ json.RawMessage) (any, error) {
	return map[string]any{"ok": true, "pluginId": r.manifest.ID}, nil
}

func (r *Runtime) capabilityCall(params json.RawMessage) (any, error) {
	var req struct {
		Name  string `json:"name"`
		Input any    `json:"input"`
	}
	if err := json.Unmarshal(params, &req); err != nil {
		return nil, err
	}
	if req.Name != "example.ping" {
		return nil, fmt.Errorf("capability not found: %s", req.Name)
	}
	return map[string]any{"pong": true, "input": req.Input}, nil
}
